x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/multi_line_string_diff_algorithm.ts

```diff
@@ -0,0 +1,114 @@
+import type {
+  ThreeVersionsOf,
+  ThreeWayDiff,
+} from '../../../../../../../../common/api/detection_engine/prebuilt_rules';
+import {
+  determineDiffOutcome,
+  determineIfValueCanUpdate,
+  ThreeWayDiffOutcome,
+  ThreeWayMergeOutcome,
+  MissingVersion,
+} from '../../../../../../../../common/api/detection_engine/prebuilt_rules';
+
+/**
+ * Diff algorithm used for string fields that contain multiple lines
+ */
+export const multiLineStringDiffAlgorithm = (
+  versions: ThreeVersionsOf<string>
+): ThreeWayDiff<string> => {
+  const {
+    base_version: baseVersion,
+    current_version: currentVersion,
+    target_version: targetVersion,
+  } = versions;
+
+  const diffOutcome = determineDiffOutcome(baseVersion, currentVersion, targetVersion);
+  const valueCanUpdate = determineIfValueCanUpdate(diffOutcome);
+
+  const { mergeOutcome, mergedVersion } = mergeVersions({
+    baseVersion,
+    currentVersion,
+    targetVersion,
+    diffOutcome,
+  });
+
+  return {
+    base_version: baseVersion,
+    current_version: currentVersion,
+    target_version: targetVersion,
+    merged_version: mergedVersion,
+
+    diff_outcome: diffOutcome,
+    merge_outcome: mergeOutcome,
+    has_update: valueCanUpdate,
+    has_conflict: mergeOutcome === ThreeWayMergeOutcome.Conflict,
+  };
+};
+
+interface MergeResult {
+  mergeOutcome: ThreeWayMergeOutcome;
+  mergedVersion: string;
+}
+
+interface MergeArgs {
+  baseVersion: string | MissingVersion;
+  currentVersion: string;
+  targetVersion: string;
+  diffOutcome: ThreeWayDiffOutcome;
+}
+
+const mergeVersions = ({
+  baseVersion,
+  currentVersion,
+  targetVersion,
+  diffOutcome,
+}: MergeArgs): MergeResult => {
+  switch (diffOutcome) {
+    case ThreeWayDiffOutcome.StockValueNoUpdate:
+    case ThreeWayDiffOutcome.CustomizedValueNoUpdate:
+    case ThreeWayDiffOutcome.CustomizedValueSameUpdate: {
+      return {
+        mergeOutcome: ThreeWayMergeOutcome.Current,
+        mergedVersion: currentVersion,
+      };
+    }
+    case ThreeWayDiffOutcome.StockValueCanUpdate: {
+      return {
+        mergeOutcome: ThreeWayMergeOutcome.Target,
+        mergedVersion: targetVersion,
+      };
+    }
+    case ThreeWayDiffOutcome.CustomizedValueCanUpdate: {
+      if (baseVersion === MissingVersion) {
+        return {
+          mergeOutcome: ThreeWayMergeOutcome.Conflict,
+          mergedVersion: currentVersion,
+        };
+      }
+      const mergedVersion = merge(currentVersion, baseVersion, targetVersion, {
+        stringSeparator: /[^\S\r\n]+/, // Separates on all whitespace except for new lines, which we keep to preserve formatting
+      });
+
+      return mergedVersion.conflict
+        ? {
+            mergeOutcome: ThreeWayMergeOutcome.Conflict,
+            mergedVersion: currentVersion,
+          }
+        : {
+            mergeOutcome: ThreeWayMergeOutcome.Merged,
+            mergedVersion: mergedVersion.result.join(' '),
```

Comment:
Not sure if it's a legitimate concern, but it looks like we are substituting all whitespace chars (except newlines) with spaces when we do `.join(' ')`. I wonder if it can mess up formatting if someone had tab chars in property values. 

I think you can't insert a tab or another weird whitespace char when editing via our web UI, but maybe you can if you edit it an external editor.

---

I played with it a little and tried to come up with a regexp that won't consume whitespace chars so that we won't have to "restore" them later. This is what I came up with:

```
stringSeparator: /(?<=\r?\n)|[^\S\r\n]+/g
```

It does split the string correctly. For example this string
```
First line.\nSecond line.
```
gets split as
```
['First', 'line.\n', 'Second', 'line.']
```
BUT for some reason `merge` still returns a conflict when I try to run it with
```
baseVersion: "First line.\nSecond line."
currentVersion: "First row.\nSecond line."
targetVersion: "First line.\nThird line."
```
I expected it merge without conflict and return
```
"First row.\nThird line."
```

@dplumlee Any idea why might that be? Am I misunderstanding how it works?

Reply:
This is a fair question, I was thinking about it when initially toying with the regex. I think the only true way around it would be to make one that gets split with all the whitespace still in tact (e.g. something like `['First ', 'line.\n', 'Second<space character>', 'line.']) so we could `.join` with no argument and just merge the strings back together. None of these fields will be impactful to rule runs themselves so I'm also unsure of it's a big concern but I suppose it has the potential to mess with markdown stuff made via third party (although by default markdown only uses spaces so far as I can tell).

For that particular example, I've been trying to get it to work for a bit - might be the case that the package is just not sophisticated enough to handle it. Or we need better regex as described above

Reply:
> I think the only true way around it would be to make one that gets split with all the whitespace still in tact (e.g. something like ['First ', 'line.\n', 'Second<space character>', 'line.']) so we could .join with no argument and just merge the strings back together with their appending whitespace attached at the hip.

Played with it a bit more and think I came up with a cleaner regexp that does exactly this: `/(?<=\s)(?=\S)/`

<img width="417" alt="Scherm­afbeelding 2024-07-15 om 19 52 58" src="https://github.com/user-attachments/assets/fb751a6d-5612-4b64-9f5a-2b3eb95d692e">


Still the `returns a computated merged version without a conflict if 3 way merge is possible` test doesn't pass. Because `merge` returns a conflict when I expect it to merge cleanly. Not sure why.
```
  {
    conflict: true,
    result: [
      'First ', 
      '<<<<<<<',
      'row.\n',  'Second ',
      '=======', 
      'line.\n', 'Third ',  
      '>>>>>>>',
      'line.'
    ]
  }
```

Do you have an idea why might that not work?

Reply:
Please disregard most of my above comment. The `/(?<=\s)(?=\S)/` regexp actually does work and tests do pass with it! I modified the test implementation, that's why I saw the error 🤦‍♂️. If you use `/(?<=\s)(?=\S)/` as `stringSeparator` and then do `.join('')` it should work fine. And this way it won't cut away non-space chars.

Also I think I found the limitation of the diffing lib. It can't merge without a conflict if changed sections are adjacent. 

For example, this merges cleanly:
```
{
  base_version: `Line\nAND\nSquare`,
  current_version: `Triangle\nAND\nSquare`,
  target_version: `Line\nAND\nCircle`,
}

// result: `Triangle\nAND\nCircle`
```

While this results in a conflict:
```
{
  base_version: `Line\nSquare`,
  current_version: `Triangle\nSquare`,
  target_version: `Line\nCircle`,
};

// result: conflict
```

But there's nothing we can do about it. Please try the new regexp approach and tell me what you think.

Reply:
Hi @dplumlee and @nikitaindik 

I've been doing some additional testing and came up with: `stringSeparator: /(\S+|\s+)/g,` plus using `.join('')`.

This allowed the algorithm to **both**:
- keep al whitespace characters
- join adjacent sections

Both of the next two examples pass:

**1.**
```ts
      const mockVersions: ThreeVersionsOf<string> = {
        base_version: `My description.\f\nThis is a second\u2001 line.\f\nThis is a third line.`,
        current_version: `My GREAT description.\f\nThis is a second\u2001 line.\f\nThis is a third line.`,
        target_version: `My description.\f\nThis is a second\u2001 line.\f\nThis is a GREAT line.`,
      };

      const expectedMergedVersion = `My GREAT description.\f\nThis is a second\u2001 line.\f\nThis is a GREAT line.`;
```

**2.**
```ts
      const mockVersions: ThreeVersionsOf<string> = {
        base_version: `Line\nSquare`,
        current_version: `Triangle\nSquare`,
        target_version: `Line\nCircle`,
      };

      const expectedMergedVersion = `Triangle\nCircle`;
```